src/include/OSL/rendererservices.h

```diff
@@ -156,6 +276,26 @@ class OSLEXECPUBLIC RendererServices {
                                   Vec3* Pout, int npoints,
                                   TypeDesc::VECSEMANTICS vectype);
 
+    /// @brief Builds a free function to provide a value for a given attribute.
+    /// @param group The shader group currently requesting the attribute.
+    /// @param object_name The object name. An empty string will be provided if
+    /// the value is not specified or it is not known at compile time.
+    /// @param attribute_name The attribute name. An empty string will be
+    /// provided if the value is not known at compile time.
+    /// @param type The type of the value being requested.
+    /// @param derivatives True if derivatives are also being requested.
+    /// @param object_lookup True if an object name was specified, even if the
+    /// value is not known at compile time.
+    /// @param array_lookup True if the attribute lookup provides an index.
+    /// @param spec The built attribute getter.
+    /// @return True if an attribute builder was provided. If false, the
+    /// existing get_attribute or get_array_attribute functions will be
+    /// used as the fallback.
+    virtual bool build_attribute_getter(ShaderGroup& group, ustring object_name,
+                                        ustring attribute_name, TypeDesc type,
+                                        bool derivatives, bool object_lookup,
+                                        bool array_lookup,
+                                        AttributeGetterSpec& spec);
```

Comment:
We might need to add one more argument to this function to allow render developers to pass through their own information.

e.g. consider a typical wrapper class:
```
class Material
{
  ShaderGroup group;

  void Compile()
  {
    shadingSystem->optimize_group(&group);
  }
};

class Renderer : RendererServices
{
  bool build_attribute_getter(ShaderGroup& group, ...) override
  {
    // how do we get access to the current Material here?
  }
};
```

Some ideas:
- Add a user supplied pointer member to `ShaderGroup`: `void Compile() { group.user_ptr = this; ... }`
- Pass a user supplied pointer through compilation: `optimize_group(this, &group);`
- Something else?

Reply:
Well, during shader execution, `ShaderGlobals::renderstate `exists so derived `RendererServices `can stow what ever it wants away.  We are suggesting a move considering this `ShaderGlobals `as `ExecContext`, meaning this is the context during shader execution.  So it figures we need a similar context during shader compilation/JIT/OPT.  Would `CompileContext `work?  Nice part is `CompileContext` is host only so free to be as complex as it wants.  Now `ExecContext `contains OSL lib side stuff and Renderer stuff as well, not sure if `CompileContext `needs to be as complex or can just be an Opaque Pointer to whatever renderer passes into optimize/jit_group.  I guess that is simplest to start with.  If so, renderer is the supplier of it, maybe `typedef void * OpaqueRendCompileContextPtr`, 
```
void optimize_group(ShaderGroup* group, OpaqueRendCompileContextPtr, ShadingContext* ctx,
                        bool do_jit = true)
```
 Although looking at that, we do have the existing host side `ShadingContext `which is already sort of a per thread thing.  Should the `ShadingContext `move towards being a `CompileContext `with a setter/getter for `RendCompileContext`.  And simply pass the `ShadingContext `through to the `bool build_attribute_getter(ShadingContext  &sc, ShaderGroup& group, ...) override`